geomstats/geometry/connection.py

```diff
@@ -41,39 +49,85 @@ def connection(self, tangent_vector_a, tangent_vector_b, base_point):
         raise NotImplementedError(
                 'connection is not implemented.')
 
-    def exp(self, tangent_vec, base_point):
+    def exp(self, tangent_vec, base_point, n_steps=N_STEPS):
         """Exponential map associated to the affine connection.
 
+        Exponential map at base_point of tangent_vec computed by integration
+        of the geodesic the equation (initial value problem), using the
+        christoffel symbols
+
         Parameters
         ----------
         tangent_vec: array-like, shape=[n_samples, dimension]
                                  or shape=[1, dimension]
 
         base_point: array-like, shape=[n_samples, dimension]
                                 or shape=[1, dimension]
-        """
-        raise NotImplementedError(
-                'The affine connection exponential is not implemented.')
 
-    def log(self, point, base_point):
+        n_steps: int
+        """
+        tangent_vec = gs.to_ndarray(tangent_vec, to_ndim=2)
+        base_point = gs.to_ndarray(base_point, to_ndim=2)
+
+        def geodesic_equation(x, v):
+            """
+            The geodesic ordinary differential equation associated
+            with the connection.
+            """
+            v = gs.to_ndarray(v, to_ndim=2)
+            gamma = self.christoffels(x)
+            return - gs.einsum('lkij,li,lj->lk', gamma, v, v)
+
+        initial_state = (base_point, tangent_vec)
+        flow, _ = self.integrate(geodesic_equation, initial_state,
+                                 n_steps=n_steps)
+        return flow[-1]
+
+    def log(self, point, base_point, n_steps=N_STEPS):
         """Logarithm map associated to the affine connection.
 
+        Solve the boundary value problem associated to the geodesic equation
+        using the christoffel symbols and an conjugate gradient descent
+
         Parameters
         ----------
         point: array-like, shape=[n_samples, dimension]
                            or shape=[1, dimension]
 
         base_point: array-like, shape=[n_samples, dimension]
                                 or shape=[1, dimension]
+
+        n_steps: int
         """
-        raise NotImplementedError(
-                'The affine connection logarithm is not implemented.')
+
+        point = gs.to_ndarray(point, to_ndim=2)
+        base_point = gs.to_ndarray(base_point, to_ndim=2)
+
+        def objective(vel):
```

Comment:
what is vel, velocity? should we call it tangent_vec to be consistent with the naming?

Reply:
I called it tangent_vec outside this subfunction